feat: allow placement of minecraft ConfiguredFeatures

dordsor21 opened the pull request:
- closes #2343
 - scope to add further features

SirYwell commented:
This only works if the EditSession is backed by an actual World, I think this should be outlined in the Javadocs.

dordsor21 commented:
noticed upstream has this functionality, will change to match them and cherry-pick where appropriate

SirYwell commented on worldedit-bukkit/adapters/adapter-1_17_1/src/main/java/com/sk89q/worldedit/bukkit/adapter/impl/fawe/v1_17_R1_2/PaperweightFaweAdapter.java:
```diff
@@ -555,6 +571,164 @@ protected ServerLevel getServerLevel(final World world) {
         return ((CraftWorld) world).getHandle();
     }
 
+    @Override
```
There is some similarity to the generateTree code here. Can we move it to the FaweAdapter too? I also think the Map<BlockPos, CraftBlockState> can be simplified to a list similar to how it's done in generateTree now.

dordsor21 replied:
I was going to look at doing so yeah. It would be more complicated though as the method changes per version a couple of times

dordsor21 replied:
Yeah I don't think that's very viable. We're not making any meaningful amount of abstraction as almost every line has a NMS class

SirYwell replied:
After looking at this again, maybe we can at least use the `getServerLevel`, `preCaptureStates`, `getCapturedBlockStatesCopy`, `postCaptureBlockStates` methods here? And also go from the Map to the List, as the Map just duplicates information but doesn't add any benefit.

dordsor21 replied:
Made the changes

SirYwell commented on worldedit-bukkit/adapters/adapter-1_20_2/src/main/java/com/sk89q/worldedit/bukkit/adapter/ext/fawe/v1_20_R2/PaperweightServerLevelDelegateProxy.java:
```diff
@@ -0,0 +1,121 @@
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+
+public class PaperweightServerLevelDelegateProxy implements InvocationHandler {
```
This class seems to be unused in that version. Did you miss upstream changes in the PaperweightAdapter? Generally, do we use that code (the changes in this PR to files in `ext/`) ourselves?

dordsor21 replied:
Yeah I think there's some cleanup required to marry the implementation across each of the adapter versions

dordsor21 replied:
Also, no, we don't use the changes in the ext classes, however, I think it's useful to have the WE code there for reference

SirYwell reviewed (changes requested):
As there's a mix of structures and features here, what exactly do we add what WE doesn't add? `StructureType` is in the WE source here, but the commands and the tools for it are in the FAWE source, that's a bit confusing to me right now.

dordsor21 commented:
> As there's a mix of structures and features here, what exactly do we add what WE doesn't add? `StructureType` is in the WE source here, but the commands and the tools for it are in the FAWE source, that's a bit confusing to me right now.

The various Tool/Factory classes are in FAWE because they are added by FAWE. The actual command methods are in com.sk89q because all commands are there

NotMyFault commented:
Please rebase for 1.20.3/4

SirYwell commented on worldedit-core/src/main/java/com/fastasyncworldedit/core/command/tool/FeaturePlacer.java:
```diff
@@ -0,0 +1,77 @@
+                boolean successful = false;
+
+                final BlockVector3 pos = clicked.toVector().add().toBlockPoint();
+                for (int i = 0; i < 10; i++) {
```
What's the repetition for?

dordsor21 replied:
I believe it exists/existed somewhere in Minecaaft or upstream? There is often a random aspect to a feature or structure (e.g. tree height) that means it might fail to be constructed occasionally. This is an arbitrary way to try a few times basically

SirYwell commented on worldedit-bukkit/adapters/adapter-1_18_2/src/main/java/com/sk89q/worldedit/bukkit/adapter/impl/fawe/v1_18_R2/PaperweightFaweAdapter.java:
```diff
@@ -547,6 +562,163 @@ protected ServerLevel getServerLevel(final World world) {
     }
 
+    @Override
+    public boolean generateFeature(ConfiguredFeatureType feature, World world, EditSession editSession, BlockVector3 pt) {
```
Could this be moved into FaweAdapter similar to generateTree?

dordsor21 replied:
See https://github.com/IntellectualSites/FastAsyncWorldEdit/pull/2364#discussion_r1403098520

SirYwell replied:
*Oh*.
I guess it's fine then, maybe I'll try to extract parts of it when working on Folia stuff, but it shouldn't be a blocker here.

dordsor21 replied:
It would just leave a method in FaweAdapter with about 5 lines in, and then the block placement separate really

SirYwell commented:
For some reason there is a folder with a `.` instead of nested folders for the 1.20.4 adapter
![dot](https://github.com/IntellectualSites/FastAsyncWorldEdit/assets/11150076/76be6cac-6825-44e2-bc0e-15558b25523b)

SirYwell commented:
From a quick test, it looks like e.g. `//structure bastion_remnant` throws an exception. I don't know if we can *fix* it, but I think it should at least be handled in a way that doesn't spam the console making it look like FAWE is doing something wrong.

Similarly, features and structures often cannot be placed with messages "Failed to generated structure. Is it a valid spot for it?" and "This feature cannot go here. Ensure the area meets the requirements." In fact, I wasn't able to spawn any feature for now. Can we somehow lift those restrictions? Otherwise, this is very limiting in its use.

dordsor21 commented:
> From a quick test, it looks like e.g. `//structure bastion_remnant` throws an exception. I don't know if we can _fix_ it, but I think it should at least be handled in a way that doesn't spam the console making it look like FAWE is doing something wrong.

Yes probably worth doing that

> Similarly, features and structures often cannot be placed with messages "Failed to generated structure. Is it a valid spot for it?" and "This feature cannot go here. Ensure the area meets the requirements." In fact, I wasn't able to spawn any feature for now. Can we somehow lift those restrictions? Otherwise, this is very limiting in its use.

I tried, but no. Minecraft generation code is complete spaghetti and it would require an incredible amount of reflection and likely need us to perform changes to the byte code. The command in upstream WE has the same limitations

SirYwell commented:
A few problems I noticed:
- e.g. `//featureplacer azalea_tree` only works when clicking on a block that can get replaced, otherwise `(FAWE) worldedit.tool.feature.failed` is printed in the chat
- `//structureplacer` defaults to none, but the tool still gets equipped and does nothing
- e.g. `//structureplacer igloo`: it seems like the structure is always placed in the same spot in every chunk, and if a chunk already has such a structure, there's no feedback.

dordsor21 commented:
For a lot of these placement issues we would potentially need to make very specific (potentially sweeping) overrides with more Unsafe forced instantiations. I'm quite sure that upstream would have the same issues - the Minecraft generation code for structures is incredibly cursed and almost every structure type would need its own implementation of fixes

SirYwell commented:
For the features (e.g. trees) it might be enough to use the block above (or depending on which side the block was clicked) for non-destructible blocks. For structures, I think we should at least try to give the user some kind of feedback if no structure was placed. Not sure if that's possible.

dordsor21 commented:
I have added the face to specific cases (most but not all, it would be even worse to add it to all that none imo). Fixed the feedback issue.

Also, for igloos etc. it's exactly the same as if the player did //set stone on the same area twice - something "happens" but we're not going to tell them nothing was changed in practice

SirYwell commented:
I still can't seem to place trees this way. Not sure if they just have weird requirements. For the face stuff, would it be easier to just try to use the face if a previous try failed?

dordsor21 commented:
> I still can't seem to place trees this way. Not sure if they just have weird requirements. For the face stuff, would it be easier to just try to use the face if a previous try failed?

We don't necessarily know why it's failed, e.g. sizes of features/structures are usually chosen before checking whether they can spawn, and thus we could end up trying to generate something we don't want to add the face to, with the face added. This is realistically the only way to do it that won't break/cause issues

dordsor21 commented:
🎉🎉
